scripts/detect-component-id-collisions.ts

```diff
@@ -0,0 +1,201 @@
+import ts from 'typescript';
+import chalk from 'chalk';
+import {readFileSync} from 'fs';
+import {join} from 'path';
+import {sync as glob} from 'glob';
+
+// This script aims to detect if the unique IDs of two components may collide at runtime
+// in order to avoid issues like https://github.com/angular/components/issues/27163.
+
+const errors: string[] = [];
+const seenMetadata = new Map<string, ts.ClassDeclaration>();
+const fileToCheck = join(__dirname, '../src/**/!(*.spec).ts');
+const ignoredPatterns = [
+  '**/components-examples/**',
+  '**/dev-app/**',
+  '**/e2e-app/**',
+  '**/universal-app/**',
+];
+
+// Use glob + createSourceFile since we don't need type information
+// and this generally faster than creating a program.
+glob(fileToCheck, {absolute: true, ignore: ignoredPatterns})
+  .map(path => ts.createSourceFile(path, readFileSync(path, 'utf8'), ts.ScriptTarget.Latest, true))
+  .forEach(sourceFile => {
+    sourceFile.statements.forEach(statement => {
+      if (ts.isClassDeclaration(statement)) {
+        checkClass(statement);
+      }
+    });
+  });
+
+if (errors.length) {
+  console.error(chalk.red('Detected identical metadata between following components:'));
+  errors.forEach(err => console.error(chalk.red(err)));
+  console.error(
+    chalk.red(
+      `\nThe metadata of one of each of these components should be ` +
+        `changed to be slightly different in order to avoid conflicts at runtime.\n`,
+    ),
+  );
+
+  process.exit(1);
+}
+
+/** Checks if a class will conflict with any of the classes that have been check so far. */
+function checkClass(node: ts.ClassDeclaration): void {
+  const metadata = getComponentMetadata(node);
+
+  if (metadata) {
+    // Create an ID for the component based on its metadata. This is based on what the framework
+    // does at runtime at https://github.com/angular/angular/blob/main/packages/core/src/render3/definition.ts#L679.
+    // Note that the behavior isn't exactly the same, because the framework uses some fields that
+    // are generated by the compiler based on the component's template.
```

Comment:
Yes, and likely we are overly sensitive here, but it guarantees we are not conflicting.